[Runtime] Handle existential target types in swift_dynamicCastMetatypeImpl

mikeash opened the pull request:
This fixes cases like `type is T.Type` when T is generic specialized to a protocol type.

Note: the compiler can still optimize these checks away and will optimize this check down to `false` in some cases. We'll want to fix that as well.

rdar://problem/56044443

mikeash committed:
[Runtime] Handle existential target types in swift_dynamicCastMetatypeImpl.

This fixes cases like `type is T.Type` when T is generic specialized to a protocol type.

Note: the compiler can still optimize these checks away and will optimize this check down to `false` in some cases. We'll want to fix that as well.

rdar://problem/56044443

DougGregor commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -1115,6 +1115,14 @@ swift_dynamicCastMetatypeImpl(const Metadata *sourceType,
+  case MetadataKind::Existential: {
+    auto targetTypeAsExistential = static_cast<const ExistentialTypeMetadata *>(targetType);
+    for (auto protocol : targetTypeAsExistential->getProtocols())
+      if (!swift_conformsToProtocol(sourceType, protocol.getSwiftProtocol()))
```
There are a couple more things to check here:
* Whether the existential is class-bounded (`isClassBounded`)
* The superclass of the existential (`getSuperclassConstraint`)
* `@objc` protocols (`getSwiftProtocol()`). You might be able to use `swift::_conformsToProtocol` rather than `swift_conformsToProtocol`, because it handles the `ProtocolDescriptorRef` directly).

mikeash replied:
Thank you, Doug! It turns out that `_conformsToProtocols` handles all of this anyway, and lets me ditch the loop. I added tests for all of these to make sure. Please let me know if you see anything else I should account for.

mikeash committed:
[Runtime] Use _conformsToProtocols instead of manually checking protocol conformance.

This is cleaner and it fixes a bunch of cases the old code didn't handle: @objc protocols, class bounds, and superclass constraints.

rdar://problem/56044443

mikeash commented:
@swift-ci plese test

DougGregor reviewed (approved):
Fantastic, thanks!

mikeash commented:
Thank you for pointing this stuff out. It would have been embarrassing to find out later.

mikeash commented:
@swift-ci please test

mikeash committed:
[Test] Split out the @objc bits of the new generic_casts.swift code.

rdar://problem/56044443

mikeash commented:
@swift-ci please test

mikeash committed:
[Test] Remove --dump-input fail from generic_casts.swift.

rdar://problem/56044443

mikeash commented:
@swift-ci please test

mikeash committed:
[Test] Make sure generic_casts.swift tests an unoptimized build of the test even when --test-optimized is passed.

rdar://problem/56044443

mikeash commented:
@swift-ci please test
